src/infer_request.cc

```diff
@@ -104,11 +105,87 @@ InferenceRequest::InferenceRequest(
     Model* model, const int64_t requested_model_version)
     : needs_normalization_(true), model_raw_(model),
       requested_model_version_(requested_model_version), flags_(0),
-      correlation_id_(0), batch_size_(0), timeout_us_(0), collect_stats_(true)
+      correlation_id_(0), batch_size_(0), timeout_us_(0), collect_stats_(true),
+      state_(InferenceRequest::State::INITIALIZED)
 {
   SetPriority(0);
 }
 
+InferenceRequest::~InferenceRequest()
+{
+  // If request has been enqueued but hasn't started executing by destruction
+  // time, an error occurred and the pending request count will need to be
+  // decremented.
+  if (state_ == InferenceRequest::State::STARTED) {
+    DecrementPendingRequestCount();
+  }
+}
+
+
+Status
+InferenceRequest::SetState(InferenceRequest::State new_state)
+{
+  // No-op if this is the current state.
+  if (state_ == new_state) {
+    return Status::Success;
+  }
+
+  // Define state transitions
+  std::stringstream ss;
+  ss << LogRequest() << "Invalid request state transition from " << state_
+     << " to " << new_state;
+  const auto error_status = Status(Status::Code::INVALID_ARG, ss.str());
+
+  switch (state_) {
+    case InferenceRequest::State::INITIALIZED: {
+      if (new_state != InferenceRequest::State::STARTED) {
+        return error_status;
+      }
+      state_ = new_state;
+      IncrementPendingRequestCount();
+      break;
+    }
+    case InferenceRequest::State::STARTED: {
+      if (new_state != InferenceRequest::State::EXECUTING) {
+        return error_status;
+      }
+      state_ = new_state;
+      DecrementPendingRequestCount();
+      break;
+    }
+    case InferenceRequest::State::EXECUTING: {
+      if (new_state != InferenceRequest::State::RELEASED) {
```

Comment:
If valid transition is linear, you may measure if the transition is correct by distance.

Reply:
Was thinking something similar. In the back of my mind I've been wondering if transitioning to `RELEASED` state from any other state should be relaxed. Ex: If some C API user creates a request and then Releases it without using it, they'd get an error (though just a log error right now). I'm wondering if it makes sense to allow Release transition from anywhere.